File: 28_KMP/KMP.cpp
/*
 * @Date: 2021-04-20 08:57:46
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-04-20 09:22:52
 */

#include <cassert>
#include <string>
#include <vector>

using namespace std;

int KMP(string haystack, string needle) {
  int n = haystack.size(), m = needle.size();
  if (m == 0) return 0;
  vector<int> pi(m);
  for (int i = 1, j = 0; i < m; i++) {
    while (j > 0 && needle[i] != needle[j]) j = pi[j - 1];
    if (needle[i] == needle[j]) j++;
    pi[i] = j;
  }

  for (int i = 0, j = 0; i < n; i++) {
    while (j > 0 && haystack[i] != needle[j]) j = pi[j - 1];
    if (haystack[i] == needle[j]) j++;
    if (j == m) return i - m + 1;
  }
  return -1;
}

int main() {
  {
    string haystack = "hello";
    string needle = "ll";
    assert(KMP(haystack, needle) == 2);
  }
  {
    string haystack = "aaaaa";
    string needle = "bba";
    assert(KMP(haystack, needle) == -1);
  }
  {
    string haystack = "";
    string needle = "";
    assert(KMP(haystack, needle) == 0);
  }
  return 0;
}

File: rust/28_KMP/KMP.rs
/*
 * @Date: 2021-04-20 08:57:51
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-04-20 09:45:14
 */

fn kmp(haystack: String, needle: String) -> i32 {
    let haystack = haystack.into_bytes();
    let needle = needle.into_bytes();
    let _h = haystack.len();
    let _n = needle.len();

    if _n == 0 {
        return 0;
    }

    let mut next = vec![0];

    for i in 1.._n {
        let mut j = next[i - 1];
        while j > 0 && needle[i] != needle[j] {
            j = next[j - 1];
        }
        next.push(if needle[i] == needle[j] { j + 1 } else { j });
    }

    let mut j = 0;

    for (i, &c) in haystack.iter().enumerate() {
        while j > 0 && c != needle[j] {
            j = next[j - 1];
        }
        if needle[j] == c {
            j += 1;
        }
        if j == _n {
            return (i + 1 - j) as i32;
        }
    }

    -1
}

fn main() {
    assert_eq!(kmp("hello".to_string(), "ll".to_string()), 2);
    assert_eq!(kmp("aaaaa".to_string(), "bba".to_string()), -1);
    assert_eq!(kmp("".to_string(), "".to_string()), 0);
}
